File: javascript/get_primes.js
"use strict";
const process = require('process');

const start = new Date();
const primes = [];
const nPrimesTarget = parseInt(process.argv[2]);
console.log(`Getting first ${nPrimesTarget} prime numbers.`);
let nPrimesFound = primes.length;

let numerator = 2;
let denominator = 2;
let limit = 0;

while (nPrimesFound < nPrimesTarget) {
  limit = Math.sqrt(numerator); 
  for (denominator = 2; denominator <= limit; denominator++) {
    if ((numerator % denominator) == 0) {
      break;
    }
  }

  // Got to here by one of two ways.
  // If here by break, then numerator is not a prime.
  // If denominator equals the numerator, then numerator is a prime.
  if (denominator > limit) {
    primes.push(numerator);
    nPrimesFound = primes.length;
  }

  numerator++;  
}

const end = new Date();
const durationMs = end - start;


let sum = 0;
let prime = 0;
primes.forEach((prime) => {
  sum += prime;
  // console.log(prime);
});

console.log(`Sum = ${sum} of first ${nPrimesFound} prime numbers.`);
console.log(`Duration: ${durationMs} [ms] (in Node)`); 

File: lua/get_primes.lua
function is_prime(n)
  result = false

  limit = math.floor(math.sqrt(n))

  numerator = n
  denominator = 2
  remainder = 0

  for denominator = 2, limit do
    remainder = math.fmod(numerator, denominator)    
    if remainder < 1 then
      break
    end
  end

  --Arrived here by one of two ways:
  --If by break, then remainder is 0, and n is not a prime.  Result is already false.
  --If remainder > 0, then the number is prime.
  if remainder > 0 then
    result = true
  end

  return result
end

--main-----------------------------------------------------------------------
function main()
  --Get target count of primes argument:
  nprimes_target = tonumber(arg[1])

  --Define early primes:
  primes = {2, 3}
  nprimes_found = #primes

  start = os.time()
  n = 5
  while nprimes_found < nprimes_target do
    prime = is_prime(n) 
    if prime then 
      table.insert(primes, n)
      nprimes_found = nprimes_found + 1
    end
  
    n = n + 2
  end

  --Sum the primes:
  sum = 0
  for i = 1, #primes do
    prime = primes[i]
    sum = sum + prime
    --print(prime)
  end

  now = os.time()
  seconds = os.difftime(now, start)
  print(string.format("The sum of the first %d prime numbers is %d", nprimes_target, sum))
  print(string.format("Duration: %d [secs] (in lua)", seconds))
end

main()
